src/server.c

```diff
@@ -3276,16 +3274,19 @@ static void propagatePendingCommands() {
  * What we want to achieve is that the entire execution unit will be done atomically,
  * currently with respect to replication and post jobs, but in the future there might
  * be other considerations. So we basically want the `postUnitOperations` to trigger
- * after the entire chain finished.
- *
- * Current, in order to avoid massive code changes that could be risky to cherry-pick,
- * we count on the mechanism we already have such as `server.core_propagation`,
- * `server.module_ctx_nesting`, and `server.in_nested_call`. We understand that we probably
- * do not need all of those variable and we will make an attempt to re-arrange it on unstable
- * branch. */
+ * after the entire chain finished. */
 void postExecutionUnitOperations() {
+    if (server.execution_nesting)
+        return;
+
     firePostExecutionUnitJobs();
+
+    /* If we are at the top-most call() and not inside a an active module
+     * context (e.g. within a module timer) we can propagate what we accumulated. */
     propagatePendingCommands();
+
+    /* Module subsystem post-execution-unit logic */
+    modulePostExecutionUnitOperations();
```

Comment:
i'd rather move this call to be explicit in the two places where module.c calls postExecutionUnitOperations.
i think that's more related to terminating the the module context, not the execution job (in that case we should maybe rename it to `moduleRootContextDestroy()`?

maybe i forgot something we discussed where you explained why that's problematic?
is it because we'll have to compare server.execution_nesting to 0? (we're also the ones who decrement it there, so it's not out of context).

Reply:
the problem is that if code from module.c runs from within call() execution_nesting will always be >0 because call() increments it.

so basically if we are in call(), the postExecutionUnitOperations that is in afterCommand actually does something (the two calls in modue.c only do something if we are outside of call)

Reply:
or we can check that the context is REDISMODULE_CTX_COMMAND and that module_ctx_nesting is **1**.
that's a bit ugly too.

i guess that's the compromise of unifying the two counters into one.
considering the two alternatives, i think the current one is better.

Reply:
we should also consider the case that cmd->proc is called directly, not through `call` (in aof.c).
in that case the dispatcher will mark the context with REDISMODULE_CTX_COMMAND and module_ctx_nesting can be either 0 (in case of startup loading) or 1 in case of DEBUG LOADAOF.

in the first case modulePostExecutionUnitOperations will not be called, and in the second case it will.
do you see any issues there?

Reply:
@oranagra for the record, this was the behavior even before this PR (CMIIW), so even if we do see an issue with that it should be handled in a separate PR

IMHO re DEBUG LOADAOF we're fine: it is a debug command, so hardly used. regardless I consider it some form of MULTI/EXEC, in which case we are aligned (PEJ will only be called after all commands were executed)
not so sure about startup AOF tho...

@MeirShpilraien did you intend for postExecutionUnitOperations to run during AOF load (both startup and DEBUG LOADAOF)

Reply:
at the very least, let's improve the tests to cover these cases, i'm paranoid about not having coverage for that and getting an assertion after releasing this change.

maybe add AOF propagation (and loading) to tests/unit/moduleapi/postnotifications.tcl
and / or maybe add restart from AOF to tests/unit/moduleapi/propagate.tcl (currently only uses DEBUG LOADAOF)

Reply:
in order to add tests, i need to know how we want and what's the expected behavior

btw, in the case of startup-AOF, won't all the PEJ accumulate and then be triggered just after the first non-AOF command?

i feel like the way to go is to prevent adding PEJs if `server.loading`

Reply:
for example, let's say a module adds some PEJ that calls `RM_Call("INCR c!", "notify-counter")` 
if we do nothing, after an AOF this counter will be 2X what it should be
on the other hand, maybe it's the module's responsibility to make things right inside the PEJ CB, based on the state.
a similar argument can be said about master-replica: the replica should not run the above RM_Call, but rather receive it from master

Reply:
i agree the modules should not do these things, and i'm not sure what these will be useful for.
i think we should call them after each command in AOF (same as we call them on a replica), maybe the module will just need to update some internal state, it should certainly not start write commands in this state.

i was paranoid about reaching some panic (even in a well behaved module), i think that our coverage for tests that do AOF loading with modules are lacking (even if they won't use the API to register PEJ, or use them to register callbacks that do nothing).

Reply:
just thought of something: if have MULTI/EXEC inside an AOF, PEJ will be executed after each command in the transaction (AOF calls `proc` but execCommand calls `call`, incrementing execution_nesting from 0 to 1 and then from 1 to 0 in `afterCommand`)

i suggest opening an issue about the correct handling of PEJ with AOF, comprised mostly of defining what we want and adding more tests. i'll mark this issue as a must for 7.2

sounds good? or do you think things should be handled in this PR?

Reply:
ok, we can leave this issue with the module PEJ to another PR, likely that RM_AddPostNotificationJob should return an error in this state.

but still, maybe in this PR we wanna improve tests/unit/moduleapi/propagate.tcl to also restart from AOF? (currently only uses DEBUG LOADAOF)